[Inductor] Optionally allow padding on non-GPU devices

blaine-rister opened the pull request:
This is the OSS component of a larger MTIA diff.
 
Currently, Inductor disables padding for non-GPU devices. We need to change this behavior to enable padding on MTIA. 

This PR adds a config option to enable padding on the CPU, or any other non-GPU device. In the future, we might want to enable padding on all devices by default. However, that might require supporting device-dependent padding defaults, since CPUs will likely use different settings than H100 GPUs.

Differential Revision: D61038114


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

shunting314 reviewed (approved):
I'm ok with this to unblock. But a cleaner way to do this is to detect we are using the MTIA backend.

blaine-rister commented:
> I'm ok with this to unblock. But a cleaner way to do this is to detect we are using the MTIA backend.

Sounds good. I'm thinking of making a follow-up PR to allow us to configure padding in the device interface. That way each device can enable/disable padding on its own. Since the MTIA backend is closed source, I didn't want OSS pytorch code to have to know about it :)

blaine-rister committed:
[Inductor-MTIA] Support padded elementwise ops (#135280)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/135280

This diff enables elementwise ops with padded tensors to run on Inductor-MTIA. This requires leveraging some new Inductor features to pad tensors to 32 byte alignment. It also updates the inductor filters to pass these shapes through, if they are only used by pointwise ops.

Currently, we're not really leveraging Inductor's reductions on MTIA, due to some triton compilation issues. So the exact details of the shape filter aren't terribly important, as the op filter is likely to reject reductions in the first place. Plus, we are planning to generalize reduction shape support in the coming weeks.

Inductor disables padding for non-GPU devices. I added a config option to enable it, so we can pad on MTIA. In the future, we might want to enable padding on all devices, and/or support device-dependent padding defaults, since CPUs (or other GPUs) will likely use different settings than H100.

Test Plan:
Added a new test to check padding e2e, and as well as a unit test to inspect the IR after fba_inductor pass. Also modified the existing tests for shape filtering.

`buck run @//mode/{opt,mtia,inplace} fbcode//mtia/compiler/graph_compiler/test_library:afg_inductor_test -- -k float16_padded `
`buck2 test @//mode/{opt,mtia,inplace} fbcode//mtia/compiler/graph_compiler/test_library:afg_inductor_unittest-artemis -- filter_input_shapes`
`buck2 test @//mode/{opt,mtia,inplace} //glow/fb/fx/fba/tests:test_fba_inductor -- padding`

Reviewed By: mrob94, jfix71

Differential Revision: D61038114
